src/decopatch/main.pyi

```diff
@@ -0,0 +1,72 @@
+from typing import Any, Callable, Optional, Protocol, TypeVar, overload
+
+from decopatch.utils_disambiguation import FirstArgDisambiguation
+from decopatch.utils_modes import SignatureInfo
+
+try:
+    from typing import ParamSpec
+except ImportError:
+    from typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+F = TypeVar("F", bound=Callable[..., Any])
+
+class _Decorator(Protocol[P]):
+    """
+    This is callable Protocol, to distinguish between cases where
+    created decorator is called as `@decorator` or `@decorator()`
+    """
+
+    # decorator is called without parenthesis: @decorator
+    @overload
+    def __call__(self, func: F) -> F: ...
+    # decorator is called with options or parenthesis: @decorator(some_option=...)
+    @overload
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> Callable[[F], F]: ...
+
+# @function_decorator is called without options or parenthesis
```

Comment:
The only downside of `pyi`, comments are placed like this, otherwise black will move to next function.

Reply:
Ok